### DISCUSSION TITLE
Possibility for multiple BT nodes in same class

### DISCUSSION THREAD

**Discussion by jonatansw** [BehaviorTree/BehaviorTree.CPP#610] [Q&A] [answered]
Hi,

First of all, thank you for this amazing library!

Regarding a ROS-based system.
The "issue" I am facing is that our autonomy engine stores a lot of information about the world the robot has created through SLAM, it's previous actions, capabilities etc., and is responsible for generating many different types of actions (generating waypoints for various tasks, safety and recovery movements, ..). As the actions use much of the same data, their implementation is in the class, and each action should have its own BT node.  
Currently we have each node as a small ros service instead, but this does not scale very well, as each new possible action from the autonomy engine will need it's own external node to communicate with.

So the question is; is there a way to have one class that offer multiple BT nodes?
I hope my question makes sense or I might have missed something in tutorials or examples, if this is at all feasible.

Thank you,
Jonatan

**Comment by facontidavide** [comment, marked as answer]
Hi,

you have two approaches!

First, check this, since it does exactly what you mention. https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/sample_nodes/crossdoor_nodes.h

On the other hand, its limitation is that these are simple nodes without Ports.

A better, but more verbose alternative, is passing the class with the shared state as a third argument, as explained in tutorial 8:

https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/examples/t08_additional_node_args.cpp
